google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadChannel.java

```diff
@@ -82,7 +98,39 @@ public void close() throws IOException {
   public int read(ByteBuffer dst) throws IOException {
     synchronized (this) {
       checkOpen();
-      int amt = channel.read(dst);
+      int amt;
+      int retries = 0;
+      int maxRetries = 3;
+      dst.mark();
+      while (true) {
+        try {
```

Comment:
Could you break up this function? It has gotten kind of long with the new logic.